pkg/terraform/provider_runner.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/utils/exec"
+)
+
+const (
+	// error messages
+	errFmtTimeout = "timed out after %v while waiting for the reattach configuration string"
+
+	// an example value would be: '{"registry.terraform.io/hashicorp/aws": {"Protocol": "grpc", "ProtocolVersion":5, "Pid":... "Addr":{"Network": "unix","String": "..."}}}'
+	envReattachConfig = "TF_REATTACH_PROVIDERS"
+	regexReattachLine = envReattachConfig + `='(.*)'`
+	reattachTimeout   = 1 * time.Minute
+)
+
+// ProviderRunner is the interface for running
+// Terraform native provider processes in the shared
+// gRPC server mode
+type ProviderRunner interface {
+	Start() (string, error)
+}
+
+// NoOpProviderRunner is a no-op ProviderRunner
+type NoOpProviderRunner struct{}
+
+// NewNoOpProviderRunner constructs a new NoOpProviderRunner
+func NewNoOpProviderRunner() NoOpProviderRunner {
+	return NoOpProviderRunner{}
+}
+
+// Start takes no action
+func (NoOpProviderRunner) Start() (string, error) {
+	return "", nil
+}
+
+// SharedProvider runs the configured native provider plugin
+// using the supplied command-line args
+type SharedProvider struct {
+	nativeProviderPath string
+	nativeProviderArgs []string
+	reattachConfig     string
+	logger             logging.Logger
+	executor           exec.Interface
+	clock              clock.Clock
+	mu                 *sync.Mutex
+}
+
+// SharedGRPCRunnerOption lets you configure the shared gRPC runner.
+type SharedGRPCRunnerOption func(runner *SharedProvider)
+
+// WithNativeProviderArgs are the arguments to be passed to the native provider
+func WithNativeProviderArgs(args ...string) SharedGRPCRunnerOption {
+	return func(sr *SharedProvider) {
+		sr.nativeProviderArgs = args
+	}
+}
+
+// WithNativeProviderExecutor sets the process executor to be used
+func WithNativeProviderExecutor(e exec.Interface) SharedGRPCRunnerOption {
+	return func(sr *SharedProvider) {
+		sr.executor = e
+	}
+}
+
+// NewSharedProvider instantiates a SharedProvider with an
+// OS executor using the supplied logger
+func NewSharedProvider(l logging.Logger, nativeProviderPath string, opts ...SharedGRPCRunnerOption) *SharedProvider {
+	sr := &SharedProvider{
+		logger:             l,
+		nativeProviderPath: nativeProviderPath,
+		executor:           exec.New(),
+		clock:              clock.RealClock{},
+		mu:                 &sync.Mutex{},
+	}
+	for _, o := range opts {
+		o(sr)
+	}
+	return sr
+}
+
+// Start starts a shared gRPC server if not already running
+// A logger, native provider's path and command-line arguments to be
+// passed to it must have been properly configured.
+// Returns any errors encountered and the reattachment configuration for
+// the native provider.
+func (sr *SharedProvider) Start() (string, error) { //nolint:gocyclo
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+	log := sr.logger.WithValues("nativeProviderPath", sr.nativeProviderPath, "nativeProviderArgs", sr.nativeProviderArgs)
```

Comment:
nitpick: we can do this initialization of `log` in `NewSharedProvider` so that it's done only once.